examples/PlaneCurve.jl

```diff
@@ -0,0 +1,273 @@
+module PlaneCurveModule
+using Oscar, Markdown
+import Base.:(==)
+
+export Point, ideal_point, AffinePlaneCurve, ProjPlaneCurve, hash, degree,
+       jacobi_ideal, curve_components, isirreducible, isreduced, reduction,
+       union
+
+################################################################################
+
+abstract type PlaneCurve end
+
+################################################################################
+# Point (not specific to curves).
+################################################################################
+# Point (first attempt).
+# The point is described by its coordinates.
+# ambient_dim gives the dimension of the space to which belongs the point.
+
+mutable struct Point{S}
+  coord::Array{S, 1}
+  ambient_dim::Int64
+  function Point(coordinates::Array{S, 1}) where {S <: FieldElem}
+    r = new{S}()
+    r.coord = coordinates
+    r.ambient_dim = length(coordinates)
+    return r
+  end
+end
+
+function Base.show(io::IO, P::Point)
+  println(io, "Point with coordinates ", P.coord)
+end
+
+################################################################################
+# Associate a maximal ideal to a point in a given ring (not specific to curves)
+
+@doc Markdown.doc"""
+    ideal_point(P::Point{S}, R::MPolyRing{S}) where S <: FieldElem
+
+Return the maximal ideal associated to the point `P` in the ring `R`.
+"""
+function ideal_point(R::MPolyRing{S}, P::Point{S}) where S <: FieldElem
+  V = gens(R)
+  return ideal(R, [V[i] - P.coord[i] for i in 1:length(V)])
+end
+
+################################################################################
+# Structure of Affine Plane Curves and Projective Plane Curves
+################################################################################
+
+mutable struct AffinePlaneCurve{S <: FieldElem} <: PlaneCurve
+  eq::Oscar.MPolyElem{S}                  # Equation of the curve (polynomial in two variables)
+  degree::Int64                           # degree of the equation of the curve
+  dimension::Int64                        # dimension of the curve (as a variety)
+  components::Dict

```

Comment:
Try to make the type concrete; my guess:
```suggestion
  components::Dict{AffinePlaneCurve{S}, Int}
```

Reply:
I don't think recursive definition like this are currently supported. To avoid the type instability, we could have a `components(..)` function to access the components, which will have the explicit type annotation.

Reply:
It seems to work for `ProjPlaneCurve`, though?

Reply:
Oh, interesting, you are right.
```julia
struct A
  x::A
end
```
does not work, but
```julia
struct A
  x::Dict{A, Int}
end
```
works. I take everythinig back.

Reply:
Oh, interesting, you are right.
```julia
struct A
  x::A
end
```
does not work, but
```julia
struct A
  x::Dict{A, Int}
end
```
works. I take everythinig back.

Reply:
It is indeed working.

Reply:
The former does not work because a non-empty struct type cannot contain another struct of the same, as then instances would require infinite storage.
But a struct type can contain *pointers* to other instances of the same struct type, and that's essentially what `Dict` does.